[NEMO-388] Off-heap memory management (reuse ByteBuffer)

hy00nc opened the pull request:
JIRA: [NEMO-388: Off-heap memory management (reuse ByteBuffer)](https://issues.apache.org/jira/projects/NEMO/issues/NEMO-388)

**Major changes:**
- New executor option: `max_offheap_mb` and `chunk_size_kb`
- `MemoryPoolAssigner` class is involved in block creation to use off-heap memory
- `ByteBuffer` is now reused, wrapped by `MemoryChunk`

**Tests for the changes:**
- `MemoryChunk` put/get tests included
- `MemoryPoolAssigner` tests included

**Other comments:**
- Design doc attached in JIRA will be edited soon

DifferentSC commented:
!rebuild

johnyangk commented:
@hy00nc let me know when I can take a look. 😄

hy00nc commented:
> @hy00nc let me know when I can take a look. 😄

It's now ready 😁(travis seems to be not working at the moment)

johnyangk commented:
Thanks for the work! I'll take a look tomorrow.

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/DirectByteBufferOutputStream.java:
```diff
@@ -106,24 +125,30 @@ public void write(final byte[] b) {
+          byteToWrite = 0;
+        }
       }
+    } catch (IllegalAccessException e) {
```
This can be removed if we remove the 'sequential' flag.
(Same goes for other similar if/else statements)

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/DirectByteBufferOutputStream.java:
```diff
@@ -106,24 +125,30 @@ public void write(final byte[] b) {
       }
+    } catch (IllegalAccessException e) {
+      throw new IOException("Not allowed for non-sequential MemoryChunk");
+    } catch (MemoryAllocationException e) {
```
I believe this is never invoked, since we never actually instantiate a 'MemoryAllocationException', at least in the current PR.
(Same goes for other similar if/else statements)

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/partition/SerializedPartition.java:
```diff
@@ -226,4 +236,11 @@ public int getLength() throws IOException {
+   * Releases the off-heap memory that this SerializedPartition holds.
+   */
+  public void release() {
+    bytesOutputStream.release();
```
Shouldn't this be called upon commit()?

write(): writes to the stream
commit(): obtains the materialized data from the stream
getData(): returns the data

Given the above, I'd think that release() operates on the data, not the stream.

hy00nc replied:
Actually, `release()` operates on the *data*, although it seems like it works on the stream. The stream here uses a list of ByteBuffers wrapped as MemoryChunk, which is the data itself. The `dataList` in SerializedPartition shares the list of ByteBuffers of the output stream, so if this is called upon commit(), it will return the 'uncleared' MemoryChunks back to the pool, causing correctness problem. Maybe I could add some comments on this?

johnyangk replied:
I'll think more about this one.

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/MemoryChunk.java:
```diff
@@ -0,0 +1,811 @@
+   *
+   * @return true, if the MemoryChunk has been freed, false otherwise.
+   */
+  public boolean isFree() {
```
Remove this unused method

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/MemoryChunk.java:
```diff
@@ -0,0 +1,811 @@
+  private final ByteBuffer buffer;
+  private long address;
+  private final long addressLimit;
+  private boolean isFree;
```
Maybe there is a better name? I thought 'free' means that this MemoryChunk has some free space to write to.

hy00nc replied:
I agree. How about `released`?

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/MemoryChunk.java:
```diff
@@ -0,0 +1,811 @@
+   *
+   * @return absolute memory address outside the heap
+   */
+  public long getAddress() {
```
Remove this unused method.

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/MemoryChunk.java:
```diff
@@ -0,0 +1,811 @@
+  protected static final long BYTE_ARRAY_BASE_OFFSET = UNSAFE.arrayBaseOffset(byte[].class);
+  private static final boolean LITTLE_ENDIAN = (ByteOrder.nativeOrder() == ByteOrder.LITTLE_ENDIAN);
+  private final ByteBuffer buffer;
+  private long address;
```
Can you briefly comment on why we access and use address/addressLimit here? (Rather than simply using the ByteBuffer abstraction)
e.g.,correctness(sanity checks), time efficiency, space efficiency, ...

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/MemoryChunk.java:
```diff
@@ -0,0 +1,811 @@
+public class MemoryChunk {
+
+  @SuppressWarnings("restriction")
+  protected static final sun.misc.Unsafe UNSAFE = getUnsafe();
```
Briefly comment on why UNSAFE is needed?

hy00nc commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/MemoryAllocationException.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * An exception to be thrown when a memory allocation operation is not successful.
+ */
+public class MemoryAllocationException extends Exception {
```
I also noticed that this exception is never thrown in this PR. I suppose it will be used in the upcoming implementation, regarding the off-heap region separation (thrown when the off-heap memory allocation exceeds the region specification). I will add the description to this exception and remove the unnecessary `catch` clauses.

hy00nc replied:
I realized I missed this part. It is supposed to be thrown while allocating direct `ByteBuffer`, because it will throw an out-of-memory error if the system lacks memory to be allocated. I will add this part so that this exception is used.

hy00nc commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/stores/BlockStore.java:
```diff
@@ -33,13 +34,14 @@
    *                             have to be handled by the scheduler with fault tolerance mechanism.)
    */
-  Block createBlock(String blockId) throws BlockWriteException;
+  Block createBlock(String blockId, MemoryPoolAssigner memoryPoolAssigner) throws BlockWriteException;
```
Injecting `MemoryPoolAssigner` directly would be really nice! Just to inform you,  every `Store` including `LocalFileStore` and `GlusterFileStore` use `MemoryPoolAssigner` while creating `FileBlock`, since it uses `SerializedPartition` that is by default written in off-heap memory.

hy00nc commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/MemoryChunk.java:
```diff
@@ -0,0 +1,811 @@
+  private final long addressLimit;
+  private boolean isFree;
+  private final int size;
+  private final boolean sequential;
```
I agree. I haven't got a concrete idea on how non-sequential MemoryChunk would be used yet. It would be nice to separate them into two subclasses to reduce confusion! I will work on that.

hy00nc replied:
I have another idea to address this suggestion: How about eliminating the `sequential` option and make `MemoryChunk`'s functions all in random mode? For the places where we use `MemoryChunk` in a sequential way, just let it access the `ByteBuffer` that `MemoryChunk` holds and use the `ByteBuffer`'s functions directly. I think this would make the codes simpler and straight forward and also avoid possible JIT compilation overhead due to subclass implementation. I will try it ASAP and share the results.

hy00nc replied:
I just pushed the changed code that is now without mode selection. I hope this isn't much confusing 😁

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/stores/AbstractBlockStore.java:
```diff
@@ -28,14 +29,18 @@
+   * @param memoryPoolAssigner the memory pool assigner.
    */
-  protected AbstractBlockStore(final SerializerManager serializerManager) {
+  protected AbstractBlockStore(final SerializerManager serializerManager,
```
Remove the `protected` modifier here and others (i.e., use no modifier). My Intellij recommends it. 

See https://stackoverflow.com/a/215505

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/stores/AbstractBlockStore.java:
```diff
@@ -48,4 +53,13 @@ protected final Serializer getSerializerFromWorker(final String blockId) {
+   *
+   * @return the memory pool assigner.
+   */
+  protected final MemoryPoolAssigner getMemoryPoolAssigner() {
```
Ditto on protected.

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/stores/AbstractBlockStore.java:
```diff
@@ -48,4 +53,13 @@ protected final Serializer getSerializerFromWorker(final String blockId) {
     final String runtimeEdgeId = RuntimeIdManager.getRuntimeEdgeIdFromBlockId(blockId);
```
Ditto on protected. (one line above)

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/stores/GlusterFileStore.java:
```diff
@@ -70,7 +73,8 @@ public Block createBlock(final String blockId) {
     final RemoteFileMetadata metadata =
       RemoteFileMetadata.create(DataUtil.blockIdToMetaFilePath(blockId, fileDirectory));
-    return new FileBlock<>(blockId, serializer, filePath, metadata);
+    final MemoryPoolAssigner memoryPoolAssigner = getMemoryPoolAssigner();
```
Very very minor suggestion (here and other FileStores): call getMemoryPoolAssigner() directly in L77 to reduce the number of variables?

return new FileBlock(..., getMemoryPoolAssigner())

johnyangk commented on runtime/executor/src/test/java/org/apache/nemo/runtime/executor/data/BlockStoreTest.java:
```diff
@@ -108,6 +108,7 @@
   private List<List<NonSerializedPartition<Integer>>> hashedBlockPartitionList;
   private List<KeyRange> readKeyRangeList;
   private List<List<Iterable>> expectedDataInRange;
+  private static final MemoryPoolAssigner memoryPoolAssigner = new MemoryPoolAssigner(3, 32);
```
Remove this unused variable.

johnyangk commented on runtime/executor/src/test/java/org/apache/nemo/runtime/executor/data/BlockTest.java:
```diff
@@ -42,6 +42,7 @@
 public final class BlockTest {
   private Serializer serializer;
   private Map<Integer, List<Integer>> testData;
+  private static final MemoryPoolAssigner memoryPoolAssigner = new MemoryPoolAssigner(1, 32);
```
This should be initialized upon setUp() (`@Before`), similar to the serializer variable.
If not, we're reusing the variable across `@Test`s, which might cause undesirable side effects.

johnyangk commented on runtime/executor/src/test/java/org/apache/nemo/runtime/executor/data/MemoryChunkTest.java:
```diff
@@ -0,0 +1,110 @@
+/**
+ * Tests basic operations on {@link MemoryChunk}.
+ */
+public class MemoryChunkTest {
```
I really like this test.

I also noticed several unused methods in MemoryChunk, getShort() for example.
Can you add tests for those methods as well?

hy00nc replied:
While refactoring MemoryChunk, we got no more use of the previous codes. They are now deleted.

johnyangk commented on runtime/executor/src/test/java/org/apache/nemo/runtime/executor/data/DirectByteBufferOutputStreamTest.java:
```diff
@@ -33,71 +34,79 @@
     int value = 1;
     outputStream.write(value);
     assertEquals(value, outputStream.toByteArray()[0]);
+    outputStream.release();
```
Seeing this made me wonder whether DirectByteBufferOutputStream#release() can be replaced with DirectByteBufferOutputStream#close(), which is empty at the moment.

I'll think more about this.

hy00nc commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/MemoryPoolAssigner.java:
```diff
@@ -0,0 +1,147 @@
+ * MemoryPoolAssigner currently supports allocation of off-heap memory only.
+ *
+ * The MemoryPoolAssigner pre-allocates all memory at the start. Memory will be occupied and reserved from start on,
+ * which means that no OutOfMemoryError comes while requesting memory. Released memory will return to the MemoryPool.
```
I forgot to update the comment on this part😅Aside from the pre-allocation, on-demand allocation is also possible. I will fix this

hy00nc commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/MemoryPoolAssigner.java:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    MemoryChunk requestChunkFromPool(final boolean sequential) throws MemoryAllocationException {
+      if (pool.isEmpty()) {
```
I will update the comment on this.

hy00nc commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/MemoryPoolAssigner.java:
```diff
@@ -0,0 +1,147 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(MemoryPoolAssigner.class.getName());
+
+  private final int chunkSize;
```
the chunk size here is translated into bytes from the chunkSizeKb.

hy00nc commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/MemoryPoolAssigner.java:
```diff
@@ -0,0 +1,147 @@
+    if (numChunks > Integer.MAX_VALUE) {
+      throw new IllegalArgumentException("Too many pages to allocate (exceeds MAX_INT)");
+    }
+    final int totalNumPages = (int) numChunks;
```
I assume this line can be removed and use `numChunks` instead of `totalNumPages`.

hy00nc replied:
I guess this line can be removed and substitute `totalNumPages` to `numChunks`

johnyangk commented:
@hy00nc I did some investigations on DirectByteBufferOutputStream#release. 

Please let me know if this makes sense. My suggestion:
- Keep DirectByteBufferOutputStream#close empty
- Remove DirectByteBufferOutputStream#release
- Change from DirectByteBufferOutputStream#getDirectByteBufferList to DirectByteBufferOutputStream#getMemoryChunkList
- In SerializedPartition#release, replace DirectByteBufferOutputStream#release with the direct invocation of memoryPoolAssigner.returnChunksToPool(dataList), and remove the offheap flag if possible. (Also using the "!committed" checker here to ensure that dataList is available would be nice)
- In DirectByteBufferOutputStreamTest, remove all invocations of DirectByteBufferOutputStream#release, and create a separate test for MemoryPoolAssigner instead if needed

================================
Here is why.

ByteArrayOutputStream, which is part of the Java API, has the following methods
- toByteArray(): returns a copy of the inner byte[]
- close(): does nothing (empty)

I think the common pattern for using this class is:
1. instantiate a new object as a local variable, and write
2. toByteArray(): obtain a copied byte[]
3. close() (explicitly or by try-with-resources)
4. the local variable goes out of scope (the inner byte[] is garbage collected)
5. The copied byte[] is garbage collected independently (perhaps after inner byte[] is garbage collected)

I think it is good to use DirectByteBufferOutputStream similarly, although our class shares the same underlying byte[] rather than copying it. This is how we're using this class at the moment.
1. Instantiate a new object as a class-level variable, and write
2. getDirectByteBufferList(): obtain ByteBufferList that points to the inner byte[]
3. close() (explicitly or by try-with-resources)
4. We now have two pointers to the same data: (1) the DirectByteBufferOutputStream variable, and (2) ByteBufferList 
5. release() on the stream
6. Both pointers now point to a 'cleared' byte[]

My issues with this pattern are:
- ByteArrayOutputStream does not provide an API similar to release() 
- The relationship between close() and release() is a little confusing
- There exist two methods that do the same thing: through releasing the stream or through directly releasing ByteBufferList
- It may be better to isolate before-commit and after-commit behaviors in SerializedPartition

Hence the above suggestion. 😄

hy00nc commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/stores/SerializedMemoryStore.java:
```diff
@@ -75,6 +79,9 @@ public void writeBlock(final Block block) throws BlockWriteException {
   @Override
   public boolean deleteBlock(final String blockId) {
+    for (Block block: getBlockMap().values()) {
+      block.release();
```
That's right I will remove the unnecessary functions. Btw, while going through this file, I noticed that although `FileBlock` and parts where `NonSerializedPartition` is converted to `SerializedPartition` use `SerializedPartition` that uses off-heap memory, there is no releasing of their off-heap memory. I will think more on this part😂

bgchun replied:
@hy00nc It'd be good to include tests that cover correct behavior of off-heap memory management. Having tests helps you consider various cases.

hy00nc replied:
That would be nice. I will try to come up with another test for this.

hy00nc commented:
> @hy00nc I did some investigations on DirectByteBufferOutputStream#release.
> 
> Please let me know if this makes sense. My suggestion:
> 
> * Keep DirectByteBufferOutputStream#close empty
> * Remove DirectByteBufferOutputStream#release
> * Change from DirectByteBufferOutputStream#getDirectByteBufferList to DirectByteBufferOutputStream#getMemoryChunkList
> * In SerializedPartition#release, replace DirectByteBufferOutputStream#release with the direct invocation of memoryPoolAssigner.returnChunksToPool(dataList), and remove the offheap flag if possible. (Also using the "!committed" checker here to ensure that dataList is available would be nice)
> * In DirectByteBufferOutputStreamTest, remove all invocations of DirectByteBufferOutputStream#release, and create a separate test for MemoryPoolAssigner instead if needed
> 
> ================================
> Here is why.
> 
> ByteArrayOutputStream, which is part of the Java API, has the following methods
> 
> * toByteArray(): returns a copy of the inner byte[]
> * close(): does nothing (empty)
> 
> I think the common pattern for using this class is:
> 
> 1. instantiate a new object as a local variable, and write
> 2. toByteArray(): obtain a copied byte[]
> 3. close() (explicitly or by try-with-resources)
> 4. the local variable goes out of scope (the inner byte[] is garbage collected)
> 5. The copied byte[] is garbage collected independently (perhaps after inner byte[] is garbage collected)
> 
> I think it is good to use DirectByteBufferOutputStream similarly, although our class shares the same underlying byte[] rather than copying it. This is how we're using this class at the moment.
> 
> 1. Instantiate a new object as a class-level variable, and write
> 2. getDirectByteBufferList(): obtain ByteBufferList that points to the inner byte[]
> 3. close() (explicitly or by try-with-resources)
> 4. We now have two pointers to the same data: (1) the DirectByteBufferOutputStream variable, and (2) ByteBufferList
> 5. release() on the stream
> 6. Both pointers now point to a 'cleared' byte[]
> 
> My issues with this pattern are:
> 
> * ByteArrayOutputStream does not provide an API similar to release()
> * The relationship between close() and release() is a little confusing
> * There exist two methods that do the same thing: through releasing the stream or through directly releasing ByteBufferList
> * It may be better to isolate before-commit and after-commit behaviors in SerializedPartition
> 
> Hence the above suggestion. 😄

Thanks for the thorough review @johnyangk ! I think the above suggestion makes the APIs much cleaner. I will refactor the code like the above. And I will also think more about how to make a useful test for `MemoryPoolAssigner`.

hy00nc commented:
So after addressing the comments, we now have two issues left to consider:

1) Whether it is worth including the code for random access of data using UNSAFE in MemoryChunk (which is not used except for tests in the project so far).

2) How to reuse (this is more focused on how *clean* the codes would be) the MemoryChunks that are used by other classes rather than SerializedMemoryBlock (i.e., NonSerializedMemoryBlock, FileBlock), since the current PR only includes reusing the MemoryChunk that belongs to SerializedMemoryBlock. 

Any suggestions/ideas would help a lot 👍

johnyangk commented:
Thanks @hy00nc.
- Regarding 1: Let's include it this time. I'll take a closer look once a new PR that uses the random access methods comes up.
- Regarding 2: It'd be good to implement this part as you suggested. 

Please let me know when you want me to take a final look at the code and merge. 😄

hy00nc commented:
> Thanks @hy00nc.
> 
> * Regarding 1: Let's include it this time. I'll take a closer look once a new PR that uses the random access methods comes up.
> * Regarding 2: It'd be good to implement this part as you suggested.
> 
> Please let me know when you want me to take a final look at the code and merge. 😄

@johnyangk thanks for the suggestions. I have implemented the reuse mechanism plz refer to this [commit](https://github.com/apache/incubator-nemo/pull/223/commits/43821610d1cdfc5529f51639ad2965ef29cc7a29). 

It will be a big help if you could review the overall fixes and edits in this PR.  Thank you😄

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/MemoryAllocationException.java:
```diff
@@ -0,0 +1,44 @@
+    super(message, cause);
+  }
+
+  public MemoryAllocationException(final String message) {
```
Other methods except this one are unused.
Perhaps the `message` argument can be removed, and something like `super("Trying to allocate memory that exceeds JobConf. MaxOffheapMb")` can be used.

hy00nc replied:
I will remove the unused methods :) And for the message, our usages include different messages like "pool empty", "lack of memory" or "exceeds MaxOffHeapMb". So maybe we should keep it?

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/MemoryPoolAssigner.java:
```diff
@@ -0,0 +1,170 @@
+      throw new IllegalArgumentException("Chunk size too small. Minimum chunk size is 4KB");
+    }
+    final long maxNumChunks = (long) maxOffheapMb * 1024 / chunkSizeKb;
+    if (maxNumChunks > Integer.MAX_VALUE) {
```
Long.MAX_VALUE?

hy00nc replied:
As we track the number of chunks in integer, this check would make sense. Also, if the number of chunks exceeds MAX_INT, it means that the memory size is set to a too large number (which wouldn't be possible).

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/MemoryPoolAssigner.java:
```diff
@@ -0,0 +1,170 @@
+      ByteBuffer buf = chunk.getBuffer();
+      buf.clear();
+      pool.add(buf);
+      chunk.release();
```
I think L159-L160 here should go inside `chunk.release()` to make things consistent. I also think `chunk.release()` should be invoked prior to `pool.add(buf)`, to ensure that a clean chunk is added back to the pool. With multi-threaded access, the chunk can be used **before** the invocation of `chunk.release()`, which may affect correctness.

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/MemoryPoolAssigner.java:
```diff
@@ -0,0 +1,170 @@
+    return chunkSize;
+  }
+
+  @VisibleForTesting
```
Remove

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/MemoryPoolAssigner.java:
```diff
@@ -0,0 +1,170 @@
+  /**
+   * Returns the number of chunks in the pool.
+   */
+  public int returnPoolSize() {
```
Remove `public`, to make it package-private.
The tests are under the same package.

poolSize()?

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/MemoryPoolAssigner.java:
```diff
@@ -0,0 +1,170 @@
+      chunk.release();
+    }
+
+    @VisibleForTesting
```
Remove

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/MemoryPoolAssigner.java:
```diff
@@ -0,0 +1,170 @@
+    }
+
+    @VisibleForTesting
+    int returnPoolSize() {
```
size()?

johnyangk commented on runtime/executor/src/test/java/org/apache/nemo/runtime/executor/data/MemoryPoolAssignerTest.java:
```diff
@@ -0,0 +1,98 @@
+
+  @After
+  public void cleanUp() {
+    this.memoryPoolAssigner = null;
```
This is not needed. (Considering other tests that we have at the moment)

johnyangk commented on runtime/executor/src/test/java/org/apache/nemo/runtime/executor/data/MemoryPoolAssignerTest.java:
```diff
@@ -0,0 +1,98 @@
+
+
+  @Test(expected = MemoryAllocationException.class)
+  public void TestTooMuchRequest() throws MemoryAllocationException {
```
testTooMuchRequest with a small 't'? 😄 
(Same for other methods)

johnyangk commented on runtime/executor/src/test/java/org/apache/nemo/runtime/executor/data/MemoryPoolAssignerTest.java:
```diff
@@ -0,0 +1,98 @@
+
+import static org.junit.Assert.assertEquals;
+
+public class MemoryPoolAssignerTest {
```
Awesome test! Thank you!

johnyangk commented on runtime/executor/src/test/java/org/apache/nemo/runtime/executor/data/MemoryPoolAssignerTest.java:
```diff
@@ -0,0 +1,98 @@
+      try{
+        chunk1 = memoryPoolAssigner.allocateChunk();
+      } catch (MemoryAllocationException e) {
+        e.printStackTrace();
```
This can go unnoticed. Throwing a RuntimeException and failing the entire test (and not only the corresponding thread) would be better.
(Same goes for all other `printStackTrace`)

johnyangk commented on runtime/executor/src/test/java/org/apache/nemo/runtime/executor/data/MemoryPoolAssignerTest.java:
```diff
@@ -0,0 +1,98 @@
+  @Test(expected = MemoryAllocationException.class)
+  public void TestConcurrentAllocation() throws MemoryAllocationException {
+    memoryPoolAssigner = new MemoryPoolAssigner(1, 32);
+    final Thread chunkThread1 = new Thread(() -> {
```
It'd be better to use an ExecutorService and also test `returnChunksToPool`.
Something like:

```java
ExecutorService e = Executors.newFixedThreadPool(5);
for (i = 0 to max) {
  e.submit(() -> {
    allocate();
    returnChunksToPool();
    allocate();
  });
}
assertEquals(max, poolSize());
```

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/partition/SerializedPartition.java:
```diff
@@ -226,4 +241,14 @@ public int getLength() throws IOException {
+  /**
+   * Releases the off-heap memory that this SerializedPartition holds.
+   */
+  public void release() {
```
What do you think about adding a JIRA issue and adding a TODO here as a pointer?

JIRA issue: Consider removing 'transient' uses of SerializedPartition. Use SerializedPartition only for SerializedMemoryStore. FileBlock and ByteOutputContext should write directly to disks/network, and not create intermediary SerializedPartitions and do extra copying. This also simplifies the uses of SerializedPartition#release().

hy00nc replied:
Thanks! I just added one.

hy00nc commented:
I addressed the requested changes. Thanks again for the suggestions and comments! I learned a lot from them👍

johnyangk commented:
@hy00nc Thanks, I'll merge once the tests pass.
